Show raster row/column of clicked point in identify results

kannes opened the pull request:
## Description
This makes identify results on raster layers include the clicked raster row/column as derived.

![grafik](https://user-images.githubusercontent.com/7661092/233475317-22837b9b-94a7-4882-8109-3ad2f8b22a80.png)

Thanks to the Value Tool plugin developers for the math <3

Implements #50421

Tested with:

- local raster TIF, GPKG and SQLite/Rasterlite = OK
- remote COG = OK
- WCS = OK
- WMS = 0, 0 <- **bad!**
- XYZ tiles are not considered by the identify tool by QGIS
- WMTS tiles are not considered by the identify tool by QGIS


# TODO
I'll try to ask around at the contributor meeting but if you read this, comments at the issue are super welcome and if you are in Den Bosch, come find me :)
- [x] Disable it for WMS layers as the results make no sense there.
- [x] ~~Any other raster layer types I should test this with?~~
- [x] Does it work with rasters that have flipped/negative axes?
- [ ] I have no idea why the "(clicked coordinate x)" and "(clicked coordinate y)" texts are written in lower case and surrounded by parenthesis. This matches vector layer identify results but even there, all other derived attributes are without parenthesis and in proper case. I much prefer that so I wrote these in that way. OK?
- [ ] I have no clue about C++ and if I used `const` and `std::floor` correctly.
- [x] Just above my additions there is a macro deciding between `derivedAttributes.unite` and `derivedAttributes.insert` depending on the Qt version, I guess I should use that too?
- [x] This should very probably be unittested I guess?

kannes committed:
Show raster row/column of clicked point in identify results

Thanks to the Value Tool plugin developers for the math <3

Implements #50421

kannes commented:
Meh, I did run `./scripts/prepare_commit.sh` but it did not complain about formatting like the failed check. I will fix it tomorrow.

jk-wg committed:
Fix indentation

nyalldawson commented on src/gui/qgsmaptoolidentify.cpp:
```diff
@@ -1054,6 +1054,18 @@ bool QgsMapToolIdentify::identifyRasterLayer( QList<IdentifyResult> *results, Qg
+  // Get raster row and column of clicked point
+  // Thanks to the Value Tool plugin developers for the math
+  const QgsRectangle extent = dprovider->extent();
+  const double xres = extent.width() / dprovider->xSize();
```
You can use layer->rasterUnitsPerPixelX() (and y) here instead.

kannes replied:
Duh, of course. Thanks :D

nyalldawson commented on src/gui/qgsmaptoolidentify.cpp:
```diff
@@ -1054,6 +1054,18 @@ bool QgsMapToolIdentify::identifyRasterLayer( QList<IdentifyResult> *results, Qg
+  const double xres = extent.width() / dprovider->xSize();
+  const double yres = extent.height() / dprovider->ySize();
+
+  const int raster_row = std::floor( ( extent.yMaximum() - point.y() ) / yres );
```
```suggestion
  const int rasterRow = std::floor( ( extent.yMaximum() - point.y() ) / yres );
```

Qgis code convention is camel case for variable names

kannes replied:
My Python brain was active, thanks.

nyalldawson commented on src/gui/qgsmaptoolidentify.cpp:
```diff
@@ -1054,6 +1054,18 @@ bool QgsMapToolIdentify::identifyRasterLayer( QList<IdentifyResult> *results, Qg
+  const int raster_row = std::floor( ( extent.yMaximum() - point.y() ) / yres );
+  const int raster_col = std::floor( ( point.x() - extent.xMinimum() ) / xres );
+
+  derivedAttributes.insert( "Clicked raster row", QLocale().toString( raster_row ) );
```
```suggestion
  derivedAttributes.insert( tr( "Clicked raster row" ), QLocale().toString( raster_row ) );
```

This string will need translation, so we wrap it in the tr(... ) function

kannes replied:
Oh good call, I forgot!

nyalldawson commented:
I've been missing this for a long time -- thanks @kannes !

jk-wg committed:
Use rasterUnitsPerPixel methods, camelcase and translation

jk-wg committed:
Add check for Qt version and use unite/insert accordingly

jk-wg committed:
Fix wrong raster pixel size being used

jk-wg committed:
Switch order so column/x is first and row/y is second

jk-wg committed:
Tests for derived raster attributes

- Change testIdentifyRaster function to return all results
- Adjust existing tests
- Add tests for derived raster attributes
- TODO fix +- half of pixel size

kannes commented:
# Issues with tests
I have added tests now but something is wrong. I meant to use logic that uses steps plus half a raster cell's size to make sure the result is identified around the center of a raster cell but I do not get the behaviour I expected. I am not sure what exactly is created in these tests and what matters. I am pretty sure this is just something stupid I am missing but I am stuck... :)

These are my issues with the current test cases (note: the raster is increasing in x left to right and decreasing in y top to bottom):

## Clicking in the zeroth column
If I'd use ` + layerXSize/2` this would click outside the raster in these two:
```
results = testIdentifyRaster( tempLayer.get(), layerExtent.xMinimum() + layerXSize, layerExtent.yMaximum() - layerYSize/2 ); // gives 0 and the min col
results = testIdentifyRaster( tempLayer.get(), layerExtent.xMinimum() + layerXSize, layerExtent.yMinimum() + layerYSize/2 );  // gives 0 and the max col
```

## Click somewhere in the middle
I would also like to use +- half the pixel size here but this would make it click on a different pixel than expected
```
const double xSomewhereCenter = layerExtent.xMinimum() + layerXSize*200 + layerXSize;  // gives 200
const double ySomewhereCenter = layerExtent.yMaximum() - layerYSize*140 - layerYSize;  // gives 140
```

nyalldawson commented on src/gui/qgsmaptoolidentify.cpp:
```diff
@@ -1054,6 +1054,20 @@ bool QgsMapToolIdentify::identifyRasterLayer( QList<IdentifyResult> *results, Qg
+  // Get raster row and column of clicked point
+  // Thanks to the Value Tool plugin developers for the math
+  const QgsRectangle extent = dprovider->extent();
+  const int rasterRow = std::floor( ( extent.yMaximum() - point.y() ) / layer->rasterUnitsPerPixelY() );
```
It's possible here that rasterUnitsPerPixel will be 0 (eg for XYZ or other rasters where the concept doesn't make sense). 

We should check this here, and only add the pixel coordinates if it's non zero

kannes replied:
Good call! I added a check.

https://github.com/qgis/QGIS/blob/66fa7e7552a762ae56bdfee1bd42f839db037cf2/src/core/raster/qgsrasterdataprovider.cpp#L305 has the same issue then I think.

jk-wg committed:
Copy logic from QgsRasterDataProvider instead of Value Tool plugin

jk-wg committed:
Protect against rasterUnitsPerPixelX/Y being zero

jk-wg committed:
Don't derive clicked column/row for dynamic image providers

jk-wg committed:
Use formatting that matches existing derived attributes

jk-wg committed:
Adjust tests

jk-wg committed:
Adjust formatting in tests

kannes commented:
Added some changes:
- I added a check for the provider type so I could skip WMS and AMS (wanted to use `dprovider->providerKey()` but that does not seem to be available). It feels super ugly to do that there but I have no clue if the providers themselves provide some info that I could check instead. Not sure if more providers need to be excluded.
- Check against raster units per pixel being zero
- Reintroduced the parenthesis around the entry in the attributes list so it matches what exists (maybe a nice issue for later to remove those globally?)

The failing compilation tests look unrelated to my code (untwine).

Still completely stumped about my tests not working as expected. I also tried to make the canvas really huge (`canvas->resize( tempLayer->width()*10, tempLayer->height()*10 );`) in case the simulated clicks were slightly off on specific canvas pixels but that changed nothing

nyalldawson commented:
Resurrected in https://github.com/qgis/QGIS/pull/54525

kannes commented:
Wooot! Thank you!
